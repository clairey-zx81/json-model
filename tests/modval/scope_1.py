#! /bin/env python
#
# Model: ./scope_1.model.json
#
# Generated by jmc (JSON Model Compiler) (https://github.com/clairey-zx81/json-model)
#

from typing import Callable
import re as re
import datetime
import urllib.parse

type Jsonable = None|bool|int|float|str|list[Jsonable]|dict[str, Jsonable]
type Path = list[str]
type Report = list[str]|None
type CheckFun = Callable[[Jsonable, str, Report], bool]
type PropMap = dict[str, CheckFun]
type TagMap = dict[None|bool|float|int|str, CheckFun]

# extract type name
def _tname(value: Jsonable) -> str:
    return type(value).__name__

# maybe add message to report
def _rep(msg: str, rep: Report) -> bool:
    rep is None or rep.append(msg)
    return False

# regex "/[a-z]/"
jm_re_0 = re.compile("[a-z]").search
# regex "/[0-9]/"
jm_re_1 = re.compile("[0-9]").search

# define "$R" ($.'$R')
def json_model_4(value: Jsonable, path: str, rep: Report = None) -> bool:
    # $.'$R'
    result = isinstance(value, list) and len(value) == 2
    if result:
        # $.'$R'.0
        result = json_model_5(value[0], path, rep)
        if not result:
            rep is None or rep.append(f"not an expected $S at {path + '.0'} [$.'$R'.0]")
        if result:
            # $.'$R'.1
            result = json_model_5(value[1], path, rep)
            if not result:
                rep is None or rep.append(f"not an expected $S at {path + '.1'} [$.'$R'.1]")
    else:
        rep is None or rep.append(f"not array or bad array length at {path} [$.'$R']")
    if not result:
        rep is None or rep.append(f"not array or unexpected array at {path} [$.'$R']")
    return result

# define "$S" ($.'$S')
def json_model_3(value: Jsonable, path: str, rep: Report = None) -> bool:
    # $.'$S'
    # "/[a-z]/"
    result = isinstance(value, str) and jm_re_0(value) is not None or _rep(f"does not match FESC at {path}", rep)
    if not result:
        rep is None or rep.append(f"not an expected REGEX at {path} [$.'$S']")
    return result

# define "$R#S" ($.'$R#S')
def json_model_5(value: Jsonable, path: str, rep: Report = None) -> bool:
    # $.'$R#S'
    # "/[0-9]/"
    result = isinstance(value, str) and jm_re_1(value) is not None or _rep(f"does not match FESC at {path}", rep)
    if not result:
        rep is None or rep.append(f"not an expected REGEX at {path} [$.'$R#S']")
    return result

# define "$" ($)
def json_model_1(value: Jsonable, path: str, rep: Report = None) -> bool:
    # $
    # $.'|'.0
    result = json_model_3(value, path, rep)
    if not result:
        rep is None or rep.append(f"not an expected $S at {path} [$.'|'.0]")
    if not result:
        # $.'|'.1
        result = json_model_5(value, path, rep)
        if not result:
            rep is None or rep.append(f"not an expected $R#S at {path} [$.'|'.1]")
    if not result:
        rep is None or rep.append(f"not any model match at {path} [$.'|']")
    return result

# entry function check_model
def check_model(value: Jsonable, path: str = "$", rep: Report = None) -> bool:
    return json_model_1(value, path, rep)


# possibly run as a script: $0 values...
if __name__ == "__main__":
    import json
    import sys
    for fn in sys.argv[1:]:
        try:
            with open(fn) as f:
                value = json.load(f)
            reasons = []
            if check_model(value, "", reasons):
                print(f"{fn}: PASS")
            else:
                print(f"{fn}: FAIL {reasons}")
        except Exception as e:
            print(f"{fn}: ERROR ({e})")

