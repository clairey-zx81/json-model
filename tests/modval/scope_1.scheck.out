scope_1.0.true.json: PASS
scope_1.1.true.json: PASS
scope_1.2.false.json: FAIL ['does not match FESC at $', "not an expected REGEX at $ [$.'$S']", "not an expected $S at $ [$.'|'.0]", 'does not match FESC at $', "not an expected REGEX at $ [$.'$R#S']", "not an expected $R#S at $ [$.'|'.1]", "not any model match at $ [$.'|']"]
scope_1.3.false.json: FAIL ['does not match FESC at $', "not an expected REGEX at $ [$.'$S']", "not an expected $S at $ [$.'|'.0]", 'does not match FESC at $', "not an expected REGEX at $ [$.'$R#S']", "not an expected $R#S at $ [$.'|'.1]", "not any model match at $ [$.'|']"]
